source/common/http/http1/codec_impl.cc

```diff
@@ -469,8 +469,41 @@ int ConnectionImpl::onHeadersCompleteBase() {
     protocol_ = Protocol::Http10;
   }
   if (Utility::isUpgrade(*current_header_map_)) {
-    ENVOY_CONN_LOG(trace, "codec entering upgrade mode.", connection_);
-    handling_upgrade_ = true;
+    // Ignore h2c upgrade requests until we support them.
+    // See https://github.com/envoyproxy/envoy/issues/7161 for details.
+    if (current_header_map_->Upgrade() &&
+        absl::EqualsIgnoreCase(current_header_map_->Upgrade()->value().getStringView(),
+                               Http::Headers::get().UpgradeValues.H2c)) {
+      ENVOY_CONN_LOG(trace, "removing unsupported h2c upgrade headers.", connection_);
+      current_header_map_->removeUpgrade();
+      if (current_header_map_->Connection()) {
+        auto values = Envoy::StringUtil::splitToken(
+            current_header_map_->Connection()->value().getStringView(), ",");
+        std::string new_values;
+        for (auto& v : values) {
+          auto value = StringUtil::toLower(StringUtil::trim(v));
+          if (value == Http::Headers::get().ConnectionValues.Upgrade) {
+            continue;
+          }
+          if (value == Http::Headers::get().ConnectionValues.Http2Settings) {
+            continue;
+          }
+          if (!new_values.empty()) {
+            new_values += ", ";
+          }
+          new_values.append(value.data(), value.size());
```

Comment:
Nit: This is probably fine, but we're now changing the case of the header values, which is a new behavior.

Reply:
I can preserve it, but I thought that envoy did that routinely.

Reply:
Ah, values, yah, fixing...

Reply:
done